### PR TITLE
main_v2 - fby3.5: cl: new sensor reading structure and support MP5990 sensor reading

### PR DESCRIPTION
**Pull request by DelphineChiu**
Summary:
- common: Phase in new sensor reading structure with README file
- fby3.5: cl: Support MP5990 sensor reading

### CODE REVIEW

**Review by GoldenBug — CHANGES_REQUESTED** [review]

File: common/dev/adm1278.c [hunk 1]
```diff
@@ -1,80 +1,202 @@
...
+
+	*val = (float)((energy_diff / sample_diff / 256) * 100) / (6123 * rsense);
+
+	return 0;
+}
+
+uint8_t adm1278_read(uint8_t sensor_num, int *reading)
+{
+	if ((reading == NULL) ||
+	    (sensor_config[sensor_config_index_map[sensor_num]].init_args == NULL)) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/adm1278.c:79]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/adm1278.c [hunk 2]
```diff
@@ -1,80 +1,202 @@
...
+	sval->fraction = (val - sval->integer) * 1000;
+	return SENSOR_READ_SUCCESS;
 }
 
-bool adm1278_read(uint8_t sensor_num, float *reading)
+uint8_t adm1278_init(uint8_t sensor_num)
 {
-	snr_cfg *cfg = &sensor_config[sensor_config_index_map[sensor_num]];
-	float val = 0;
+	if (!sensor_config[sensor_config_index_map[sensor_num]].init_args) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/adm1278.c:160]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/adm1278.c [hunk 3]
```diff
@@ -1,80 +1,202 @@
...
+	if (sample_diff == 0) {
+		return -1;
+	}
+
+	*val = (float)((energy_diff / sample_diff / 256) * 100) / (6123 * rsense);
+
+	return 0;
+}
+
+uint8_t adm1278_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/adm1278.c:76]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/adm1278.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/adm1278.c:160]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/ast_adc.c [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		printf("<error> ADC[%d] is invalid device!\n", index);
+		return false;
+	}
+
+	if (!is_ready[index]) {
+		printf("<error> ADC[%d] is not ready to read!\n", index);
+		return false;
+	}
+
+	int err, retval;
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/ast_adc.c]: We have fixed this and do the forced update.
Please review it. Thanks.

File: common/dev/ast_adc.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	sensor_val *sval = (sensor_val *)reading;
+	sval->integer = (val / 1000) & 0xFFFF;
+	sval->fraction = (val % 1000) & 0xFFFF;
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t ast_adc_init(uint8_t sensor_num)
+{
+	if (!sensor_config[sensor_config_index_map[sensor_num]].init_args) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/ast_adc.c:137]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/intel_peci.c [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+	if (!reading) {
+		return false;
+	}
+
+	const uint16_t param = 0x00;
+	const uint8_t rlen = 0x05;
+	uint8_t rbuf[rlen];
+	memset(rbuf, 0, sizeof(rbuf));
+
+	if (peci_retry_read(PECI_CMD_RD_PKG_CFG0, addr, RDPKG_IDX_TJMAX_TEMP, param, rlen, rbuf) ==
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/intel_peci.c]: We have fixed this and do the forced update.
Please review it. Thanks.

File: common/dev/intel_peci.c [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+	if (!reading) {
+		return false;
+	}
+
+	const uint16_t param = 0xFF;
+	const uint8_t rlen = 0x05;
+	uint8_t rbuf[rlen];
+	memset(rbuf, 0, sizeof(rbuf));
+
+	if (peci_retry_read(PECI_CMD_RD_PKG_CFG0, addr, RDPKG_IDX_PKG_TEMP, param, rlen, rbuf) ==
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/intel_peci.c]: We have fixed this and do the forced update.
Please review it. Thanks.

File: common/dev/isl28022.c [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+#define ISL28022_CONFIG_REG 0
+#define ISL28022_BUS_VOLTAGE_REG 2
+#define ISL28022_POWER_REG 3
+#define ISL28022_CURRENT_REG 4
+#define ISL28022_CALIBRATION_REG 5
+
+uint8_t isl28022_read(uint8_t sensor_num, int *reading)
+{
+	if ((reading == NULL) ||
+	    (sensor_config[sensor_config_index_map[sensor_num]].init_args == NULL)) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/isl28022.c:15]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/isl28022.c [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	} else {
+		return SENSOR_FAIL_TO_ACCESS;
+	}
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t isl28022_init(uint8_t sensor_num)
+{
+	if (sensor_config[sensor_config_index_map[sensor_num]].init_args == NULL) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/isl28022.c:80]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/isl69259.c [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	if (reading == NULL) {
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	uint8_t retry = 5;
+	sensor_val *sval = (sensor_val *)reading;
+	I2C_MSG msg;
+	memset(sval, 0, sizeof(sensor_val));
+
+	msg.bus = sensor_config[sensor_config_index_map[sensor_num]].port;
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/isl69259.c:18]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/ltc4282.c [hunk 11]
```diff
@@ -0,0 +1,65 @@
+#include <stdio.h>
+#include <string.h>
+#include "sensor.h"
+#include "hal_i2c.h"
+
+uint8_t ltc4282_read(uint8_t sensor_num, int *reading)
+{
+	if ((reading == NULL) ||
+	    (sensor_config[sensor_config_index_map[sensor_num]].init_args == NULL)) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/ltc4282.c:9]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/ltc4282.c [hunk 12]
```diff
@@ -0,0 +1,65 @@
+#include <stdio.h>
+#include <string.h>
+#include "sensor.h"
+#include "hal_i2c.h"
+
+uint8_t ltc4282_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/ltc4282.c:6]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/isl28022.c [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+		sval->integer = read_power;
+		sval->fraction = ((read_power - sval->integer) * 1000);
+	} else {
+		return SENSOR_FAIL_TO_ACCESS;
+	}
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t isl28022_init(uint8_t sensor_num)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/isl28022.c:74]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/isl28022.c [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+#include "sensor.h"
+#include "hal_i2c.h"
+
+#define ISL28022_CONFIG_REG 0
+#define ISL28022_BUS_VOLTAGE_REG 2
+#define ISL28022_POWER_REG 3
+#define ISL28022_CURRENT_REG 4
+#define ISL28022_CALIBRATION_REG 5
+
+uint8_t isl28022_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/isl28022.c:12]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/ltc4282.c [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+	sensor_val *sval = (sensor_val *)reading;
+	memset(sval, 0, sizeof(*sval));
+
+	sval->integer = (int)val & 0xFFFF;
+	sval->fraction = (val - sval->integer) * 1000;
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t ltc4282_init(uint8_t sensor_num)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/ltc4282.c:61]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/mp5990.c [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+#include <string.h>
+#include <stdlib.h>
+#include "sensor.h"
+#include "hal_i2c.h"
+#include "libutil.h"
+#include "pmbus.h"
+
+#define I2C_DATA_SIZE 5
+
+uint8_t mp5990_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/mp5990.c:11]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/mp5990.c [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+	sensor_val *sval = (sensor_val *)reading;
+	memset(sval, 0, sizeof(*sval));
+
+	sval->integer = (int32_t)val;
+	sval->fraction = (int32_t)(val * 1000) % 1000;
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t mp5990_init(uint8_t sensor_num)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/mp5990.c:72]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/mp5990.c [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+#include "hal_i2c.h"
+#include "libutil.h"
+#include "pmbus.h"
+
+#define I2C_DATA_SIZE 5
+
+uint8_t mp5990_read(uint8_t sensor_num, int *reading)
+{
+	if ((reading == NULL) ||
+	    (sensor_config[sensor_config_index_map[sensor_num]].init_args == NULL)) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/mp5990.c:14]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/mp5990.c [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+	sval->integer = (int32_t)val;
+	sval->fraction = (int32_t)(val * 1000) % 1000;
+
+	return SENSOR_READ_SUCCESS;
+}
+
+uint8_t mp5990_init(uint8_t sensor_num)
+{
+	if (!sensor_config[sensor_config_index_map[sensor_num]].init_args) {
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/mp5990.c:78]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/nvme.c [hunk 20]
```diff
@@ -1,78 +1,50 @@
...
 
-#define RETRY 5
+#define NVMe_NOT_AVAILABLE 0x80
+#define NVMe_TMP_SENSOR_FAILURE 0x81
 
-#define NVME_NOT_AVAILABLE 0x80
-#define NVME_TEMP_SENSOR_FAILURE 0x81
-
-bool nvme_read(uint8_t sensor_num, float *reading)
+uint8_t nvme_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/nvme.c:8]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/pch.c [hunk 21]
```diff
@@ -1,59 +1,50 @@
...
 #include <stdlib.h>
-
-#include "ipmi.h"
 #include "sensor.h"
-#include "plat_sensor_table.h"
+#include "libutil.h"
+#include "ipmi.h"
 
-bool pch_read(uint8_t sensor_num, float *reading)
+uint8_t pch_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/pch.c:7]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/pex89000.c [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+		float tmp = (resp & 0xFFFF) / 128;
+		val->integer = (int)tmp;
+		val->fraction = (int)(tmp * 1000) % 1000;
+		return true;
+	}
+
+	return false;
+}
+
+uint8_t pex89000_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/pex89000.c:298]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/tmp75.c [hunk 23]
```diff
@@ -1,41 +1,32 @@
...
-
 #include <stdio.h>
-
 #include "sensor.h"
 #include "hal_i2c.h"
 
-#define RETRY 5
-
-bool tmp75_read(uint8_t sensor_num, float *reading)
+uint8_t tmp75_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/tmp75.c:5]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/tps53689.c [hunk 24]
```diff
@@ -0,0 +1,56 @@
+#include <stdio.h>
+#include <string.h>
+#include "sensor.h"
+#include "hal_i2c.h"
+#include "pmbus.h"
+#include "util_pmbus.h"
+
+uint8_t tps53689_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/tps53689.c:8]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/dev/xdpe15284.c [hunk 25]
```diff
@@ -0,0 +1,62 @@
+#include <stdio.h>
+#include <string.h>
+#include "sensor.h"
+#include "hal_i2c.h"
+#include "util_pmbus.h"
+#include "pmbus.h"
+
+uint8_t xdpe15284_read(uint8_t sensor_num, int *reading)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/xdpe15284.c:8]: We have fixed this and do the forced update.
The sensor_num is unsigned type so only judge if it's less than MAX_NUM_SENSORS.
Please review it. Thanks.

File: common/lib/util_pmbus.c [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+	/* Sign extend Mantissa to 16 bits */
+	if (mantissa > 0x03FF)
+		mantissa |= 0xF800;
+
+	lsb = slinear11_exponents[exponent];
+	ret = mantissa * lsb;
+	return ret;
+}
+
+bool get_exponent_from_vout_mode(uint8_t sensor_num, float *exponent)
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/lib/util_pmbus.c:57]: We have fixed this and do the forced update.
Please review it. Thanks.

File: common/dev/nvme.c [hunk 27]
```diff
@@ -1,78 +1,50 @@
...
-		}
-		not_available_retry_num = 0;
-	} else {
-		cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
-		return false;
-	}
+		if (val == NVMe_NOT_AVAILABLE)
+			return SENSOR_FAIL_TO_ACCESS;
+		else if (val == NVMe_TMP_SENSOR_FAILURE)
+			return SENSOR_UNSPECIFIED_ERROR;
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/nvme.c:37]: We have fixed this and do the forced update.
Please review it. Thanks.

File: common/dev/nvme.c [hunk 28]
```diff
@@ -1,78 +1,50 @@
...
+	} else
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/nvme.c]: We have fixed this and do the forced update.
Please review it. Thanks.

File: common/dev/pex89000.c [hunk 29]
```diff
@@ -0,0 +1,359 @@
...
+{
+	if (reading == NULL) {
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	pex89000_init_arg *init_arg =
+		(pex89000_init_arg *)sensor_config[sensor_config_index_map[sensor_num]].init_args;
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/pex89000.c:305]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/tmp75.c [hunk 30]
```diff
@@ -1,41 +1,32 @@
...
+	if (!reading) {
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	uint8_t retry = 5;
+	I2C_MSG msg = { 0 };
 
-	msg.bus = cfg->port;
-	msg.target_addr = cfg->target_addr;
+	msg.bus = sensor_config[sensor_config_index_map[sensor_num]].port;
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/tmp75.c:14]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/tps53689.c [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+	if (reading == NULL) {
+		return SENSOR_UNSPECIFIED_ERROR;
+	}
+
+	uint8_t retry = 5;
+	sensor_val *sval = (sensor_val *)reading;
+	I2C_MSG msg;
+	memset(sval, 0, sizeof(sensor_val));
+
+	msg.bus = sensor_config[sensor_config_index_map[sensor_num]].port;
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/tps53689.c:19]: The stack memory is limited to BIC.
The size of sensor_cfg is 60 bytes.
So, we propose not to define a temporary variable.

File: common/dev/pex89000.c [hunk 32]
```diff
@@ -0,0 +1,359 @@
...
+			k_msleep(10);
+			continue;
+		}
+
+		if ((resp >> 24) & 0x8) { // CHIME_to_AXI_CSR Control Status -> Read_data_vaild
+			return true; //  success
+		}
+
+		k_msleep(10);
+	}
```
[the remark this answers is not in the document]
Comment (RenChen-wiwynn) [reply common/dev/pex89000.c]: We have fixed this and do the forced update.
Please review it. Thanks.

**Comment by DelphineChiu** [comment]
close the PR since the codes have been merged

**Pull request closed without merging**